Email stats: Add user-content-links to email stats

TimBroddin opened the pull request:
Closes #81881 

## Proposed Changes

In the previous implementation, we only tracked internal links for email statistics. However, there is also a need to include and track user-content-links.

This PR adds user content links to the links overview in Calypso:

![CleanShot 2023-09-19 at 10 46 01@2x](https://github.com/Automattic/wp-calypso/assets/528287/ffabc3b3-2951-45aa-8dce-a2d4139d864d)


## Testing Instructions

Since the user content link tracking isn't live yet, the easiest way to test this is to modify the endpoint on wpcom to provide mock data: fbhepr%2Skers%2Sjcpbz%2Schoyvp.ncv%2Serfg%2Sjcpbz%2Qwfba%2Qraqcbvagf%2Spynff.jcpbz%2Qwfba%2Qncv%2Qfgngf%2Qrznvy%2Qpyvpxf%2Qol%2Qhfre%2Qpbagrag%2Qyvax%2Qi1%2Q1%2Qraqcbvag.cuc%2322%2Q22-og 

Add this on line 10:

```php

		return array( 'user-content-links' => array('data' => array(
			['https://google.com', 10],
			['https://youtube.com', 200]
		), 'fields' => array( 'url', 'clicks_count' ) ) );
```

Then:
1. Apply this PR
2. Go to Stats & click an email inside the `Emails ` box
3. Click `Email clicks` in the navigation & you should see the user content links inside the Links box
4. Clicking a link will open a new tab with that link./

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

TimBroddin committed:
Add user-content-links to email stats

In the previous implementation, we only tracked internal links for email statistics. However, there is also a need to include and track user-content-links.

The changes introduced in this commit ensure that user-content-links are also fetched along with ‘link’ stats. Thereafter, these user-content-links are parsed and included in the overall email stats. The code now includes checks for the existence of both internalLinks and userContentLinks. It also includes the sorting of the 'mappedLinks' array..

JuanLucha commented on client/state/stats/emails/utils.js:
```diff
@@ -218,12 +218,13 @@ export function parseEmailListData( list ) {
-	if ( ! links ) {
-		return null;
+export function parseEmailLinksData( internalLinks, userContentLinks ) {
+	if ( ! Array.isArray( internalLinks ) || ! Array.isArray( userContentLinks ) ) {
```
According to the call to this function:
```js
		links: parseEmailLinksData( stats.links?.data, stats[ 'user-content-links' ]?.data ),
```
The second parameter could be empty. In that case, the internalLinks won't be parsed.

TimBroddin replied:
You're 100% right, I'll refactor

JuanLucha commented on client/state/stats/emails/utils.js:
```diff
@@ -234,30 +235,40 @@ export function parseEmailLinksData( links ) {
 
-	return mappedLinks;
+	// add user content links
+	userContentLinks.forEach( ( link ) => {
```
If we do something about the previous comment we need to check userContentLinks is array here.

TimBroddin committed:
Refactor parseEmailLinksData for array validation

The function parseEmailLinksData has been updated to provide default values for the input parameters. This change removes the need for an early function exit when the input parameters are not arrays. The validation check for the input parameters has been moved for each variable separately to correctly handle the scenario when one of the parameters is not an array. This refactor greatly simplifies the function logic and improves its robustness.
